File: src/api/storage/cache.js
/**
 * 缓存存储 API — 纯内存，支持 TTL 过期
 *
 * 数据仅存在于进程生命周期内，适合临时缓存、最近播放等场景。
 *
 * @example
 * import { cacheGet, cacheSet, cacheTtl } from '@/api/storage/cache';
 * // 缓存 10 分钟
 * await cacheSet('recent_tracks', data, cacheTtl.durationSecs(600));
 * // 永不过期
 * await cacheSet('app_state', data, cacheTtl.forever);
 */

import { transport } from '@/api/transport';
import { perf } from '@/utils/performanceMonitor.js';

// ══════════════════════════════════════════════════════════════════════════════
// TTL 辅助工具
// ══════════════════════════════════════════════════════════════════════════════

/**
 * TTL 策略工厂
 *
 * @example
 * cacheTtl.forever            // → "forever"
 * cacheTtl.session            // → "session"
 * cacheTtl.durationSecs(600)  // → { duration_secs: 600 }
 */
export const cacheTtl = {
  /** 永不过期 */
  get forever() {
    return 'forever';
  },
  /** 当前进程生命周期 */
  get session() {
    return 'session';
  },
  /** 指定秒数后过期 */
  durationSecs(secs) {
    return { duration_secs: secs };
  },
};

// ══════════════════════════════════════════════════════════════════════════════
// Cache API
// ══════════════════════════════════════════════════════════════════════════════

/**
 * 读取缓存值（自动检查过期）
 * @param {string} key - 缓存键
 * @returns {Promise<any>} 缓存值
 */
export async function cacheGet(key) {
  return perf.measureAsync('cache.get', transport.command('cache_get', { key }), { key });
}

/**
 * 写入缓存值并指定 TTL
 * @param {string} key - 缓存键
 * @param {any} value - 要缓存的值
 * @param {string|{duration_secs: number}} ttl - TTL 策略，使用 cacheTtl 辅助对象
 * @returns {Promise<void>}
 *
 * @example
 * await cacheSet('recent', data, cacheTtl.durationSecs(600));
 * await cacheSet('state', data, cacheTtl.forever);
 */
export async function cacheSet(key, value, ttl) {
  return perf.measureAsync('cache.set', transport.command('cache_set', { key, value, ttl }), { key });
}

/**
 * 删除缓存项
 * @param {string} key - 缓存键
 * @returns {Promise<boolean>} 是否成功删除
 */
export async function cacheRemove(key) {
  return perf.measureAsync('cache.remove', transport.command('cache_remove', { key }), { key });
}

/**
 * 检查缓存项是否存在且未过期
 * @param {string} key - 缓存键
 * @returns {Promise<boolean>}
 */
export async function cacheHas(key) {
  return perf.measureAsync('cache.has', transport.command('cache_has', { key }), { key });
}

/**
 * 获取所有未过期的缓存键
 * @returns {Promise<string[]>}
 */
export async function cacheKeys() {
  return perf.measureAsync('cache.keys', transport.command('cache_keys'));
}

/**
 * 清空所有缓存（含未过期条目）
 * @returns {Promise<void>}
 */
export async function cacheClear() {
  return perf.measureAsync('cache.clear', transport.command('cache_clear'));
}

/**
 * 清理所有已过期条目
 * @returns {Promise<number>} 清理数量
 */
export async function cacheClearExpired() {
  return perf.measureAsync('cache.clearExpired', transport.command('cache_clear_expired'));
}

/**
 * 续期缓存项：按给定 TTL 重置过期倒计时
 * @param {string} key - 缓存键
 * @param {string|{duration_secs: number}} ttl - 新的 TTL 策略
 * @returns {Promise<boolean>} 是否续期成功
 */
export async function cacheTouch(key, ttl) {
  return perf.measureAsync('cache.touch', transport.command('cache_touch', { key, ttl }), { key });
}

File: src-tauri/crates/chordial-server/src/routes/cache.rs
//! Cache 域路由 — 内存 TTL 缓存（含 Blob 磁盘存储）。
//!
//! | 方法 | 路径 | 对应命令 |
//! |------|------|---------|
//! | GET | `/cache/:key` | `cache_get` |
//! | PUT | `/cache/:key` | `cache_set` (body 含 ttl) |
//! | DELETE | `/cache/:key` | `cache_remove` |
//! | HEAD | `/cache/:key` | `cache_has` |
//! | GET | `/cache` | `cache_keys` |
//! | DELETE | `/cache` | `cache_clear` |
//! | POST | `/cache/clear-expired` | `cache_clear_expired` |
//! | POST | `/cache/touch` | `cache_touch` |
//! | POST | `/cache/blob/enable` | `cache_enable_blob_storage` |
//! | GET | `/cache/blob/enabled` | `cache_blob_storage_enabled` |
//! | PUT | `/cache/blob/:key` | `cache_set_blob` (body: bytes + ttl header) |
//! | GET | `/cache/blob/:key` | `cache_get_blob` |
//! | DELETE | `/cache/blob/:key` | `cache_remove_blob` |
//! | HEAD | `/cache/blob/:key` | `cache_has_blob` |
//! | GET | `/cache/blob` | `cache_blob_keys` |
//! | DELETE | `/cache/blob` | `cache_clear_blobs` |
//! | POST | `/cache/blob/clear-expired` | `cache_clear_expired_blobs` |

use crate::state::AppState;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::routing::{get, post};
use axum::{Json, Router};
use chordial_core::module::storage::entry::Ttl;
use serde::Deserialize;
use serde_json::Value;

/// 带 TTL 的缓存设置请求体。
#[derive(Debug, Deserialize)]
struct CacheSetBody {
    value: Value,
    #[serde(default)]
    ttl: TtlArg,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
enum TtlArg {
    Forever,
    Session,
    DurationSecs(u64),
}

impl From<TtlArg> for Ttl {
    fn from(arg: TtlArg) -> Self {
        match arg {
            TtlArg::Forever => Ttl::Forever,
            TtlArg::Session => Ttl::Session,
            TtlArg::DurationSecs(n) => Ttl::DurationSecs(n),
        }
    }
}

impl Default for TtlArg {
    fn default() -> Self {
        TtlArg::Session
    }
}

pub fn router() -> Router<AppState> {
    Router::new()
        // 纯内存缓存
        .route("/cache", get(keys).delete(clear))
        .route("/cache/clear-expired", post(clear_expired))
        .route("/cache/touch", post(touch))
        .route("/cache/:key", get(get_one).put(set).delete(remove).head(has))
        // Blob 缓存
        .route("/cache/blob/enable", post(enable_blob))
        .route("/cache/blob/enabled", get(blob_enabled))
        .route("/cache/blob", get(blob_keys).delete(clear_blobs))
        .route("/cache/blob/clear-expired", post(clear_expired_blobs))
        .route(
            "/cache/blob/:key",
            get(blob_get).put(blob_set).delete(blob_remove).head(blob_has),
        )
}

// ── 纯内存 ──────────────────────────────────────────

async fn get_one(State(state): State<AppState>, Path(key): Path<String>) -> impl IntoResponse {
    match state.ctx.cache.get_raw(&key) {
        Some(v) => Ok(Json(v)),
        None => Err((StatusCode::NOT_FOUND, format!("缓存项 '{}' 不存在或已过期", key))),
    }
}

async fn set(
    State(state): State<AppState>,
    Path(key): Path<String>,
    Json(body): Json<CacheSetBody>,
) -> StatusCode {
    state.ctx.cache.set_raw(&key, body.value, &body.ttl.into());
    StatusCode::NO_CONTENT
}

async fn remove(State(state): State<AppState>, Path(key): Path<String>) -> Json<bool> {
    Json(state.ctx.cache.remove(&key))
}

async fn has(State(state): State<AppState>, Path(key): Path<String>) -> StatusCode {
    if state.ctx.cache.has(&key) {
        StatusCode::OK
    } else {
        StatusCode::NOT_FOUND
    }
}

async fn keys(State(state): State<AppState>) -> Json<Vec<String>> {
    Json(state.ctx.cache.keys())
}

async fn clear(State(state): State<AppState>) -> StatusCode {
    state.ctx.cache.clear();
    StatusCode::NO_CONTENT
}

async fn clear_expired(State(state): State<AppState>) -> Json<usize> {
    Json(state.ctx.cache.clear_expired())
}

/// `POST /cache/touch` — 续期
/// body: `{ "key": "...", "ttl": "session" }`
#[derive(Debug, Deserialize)]
struct TouchBody {
    key: String,
    ttl: TtlArg,
}

async fn touch(
    State(state): State<AppState>,
    Json(body): Json<TouchBody>,
) -> Json<bool> {
    Json(state.ctx.cache.touch(&body.key, &body.ttl.into()))
}

// ── Blob ────────────────────────────────────────────

/// `POST /cache/blob/enable` — 启用 Blob 磁盘存储
/// body: `{ "dir": "/path/to/blobs" }`
#[derive(Debug, Deserialize)]
struct EnableBlobBody {
    dir: String,
}

async fn enable_blob(
    State(state): State<AppState>,
    Json(body): Json<EnableBlobBody>,
) -> Result<StatusCode, String> {
    state.ctx.cache.enable_blob_storage(std::path::PathBuf::from(body.dir))?;
    Ok(StatusCode::NO_CONTENT)
}

async fn blob_enabled(State(state): State<AppState>) -> Json<bool> {
    Json(state.ctx.cache.blob_storage_enabled())
}

/// Blob 设置请求体。
#[derive(Debug, Deserialize)]
struct BlobSetBody {
    data: String, // base64 encoded
    #[serde(default)]
    ttl: TtlArg,
}

async fn blob_set(
    State(state): State<AppState>,
    Path(key): Path<String>,
    Json(body): Json<BlobSetBody>,
) -> Result<StatusCode, String> {
    use base64::Engine;
    let data = base64::engine::general_purpose::STANDARD
        .decode(&body.data)
        .map_err(|e| format!("base64 解码失败: {}", e))?;
    state.ctx.cache.set_blob(&key, &data, &body.ttl.into())?;
    Ok(StatusCode::NO_CONTENT)
}

async fn blob_get(State(state): State<AppState>, Path(key): Path<String>) -> impl IntoResponse {
    match state.ctx.cache.get_blob(&key) {
        Some(data) => Ok::<_, (StatusCode, String)>(data),
        None => Err((StatusCode::NOT_FOUND, format!("Blob 缓存项 '{}' 不存在或已过期", key))),
    }
}

async fn blob_remove(State(state): State<AppState>, Path(key): Path<String>) -> Json<bool> {
    Json(state.ctx.cache.remove_blob(&key))
}

async fn blob_has(State(state): State<AppState>, Path(key): Path<String>) -> StatusCode {
    if state.ctx.cache.has_blob(&key) {
        StatusCode::OK
    } else {
        StatusCode::NOT_FOUND
    }
}

async fn blob_keys(State(state): State<AppState>) -> Json<Vec<String>> {
    Json(state.ctx.cache.blob_keys())
}

async fn clear_blobs(State(state): State<AppState>) -> StatusCode {
    state.ctx.cache.clear_blobs();
    StatusCode::NO_CONTENT
}

async fn clear_expired_blobs(State(state): State<AppState>) -> Json<usize> {
    Json(state.ctx.cache.clear_expired_blobs())
}
